src/bio.c

```diff
@@ -345,14 +351,57 @@ int wolfSSL_BIO_read(WOLFSSL_BIO* bio, void* buf, int len)
             #ifdef USE_WOLFSSL_IO
                 /* BIO requires built-in socket support
                  *  (cannot be used with WOLFSSL_USER_IO) */
+                bio->flags &= ~WOLFSSL_BIO_FLAG_RETRY;
                 ret = wolfIO_Recv(bio->num, (char*)buf, len, 0);
+                if (ret < 0) {
+#ifdef USE_WINDOWS_API
+                    if (WSAGetLastError() == WSAEWOULDBLOCK)
+                        bio->flags |= WOLFSSL_BIO_FLAG_RETRY;
+#else
+                    if (errno == EAGAIN)
+                        bio->flags |= WOLFSSL_BIO_FLAG_RETRY;
+#endif
+                }
```

Comment:
Please leave system dependent stuff inside `wolfio.c`. This will become a mess handling it here. I would recommend `wolfIO_RecvFrom` and `wolfIO_SendTo` use `TranslateIoError` instead of `TranslateReturnCode`. You can change their names so that they are less similar to `wolfIO_Recv` and `wolfIO_Send`.

Reply:
Fully refactored.  I leveraged existing routines in wolfio.c to make this easier and more portable.  I'm now communicating the `EAGAIN`/`EWOULDBLOCK`/`ETIMEDOUT` to bio.c using wolfssl-native `SOCKET_NODATA`, which are remapped to `WOLFSSL_BIO_ERROR` by `wolfSSL_BIO_read()` and `wolfSSL_BIO_write()`, with appropriate `bio->flags` update.